scripts/04-find-patterns.py

```diff
@@ -20,73 +20,86 @@
-writer = csv.writer(
-    csv_file, delimiter=';',
-    quotechar='"', quoting=csv.QUOTE_MINIMAL)
-writer.writerow(['filename', 'var_middle_number', 'nested_double_for_number'])
-
 
-def log_result(result):
+def log_result(result, file_to_write):
     """ Save result to csv file. """
-    # print(len(result[1]))
-    writer.writerow([result[0], len(result[1]), len(result[2])])
-    csv_file.flush()
+    with open(file_to_write, 'a', newline='\n', encoding='utf-8') as csv_file:
+        p = Path(result[0])
+        d_path = Path(dir_path)
+        relative_path = p.relative_to(d_path)
 
+        writer = csv.writer(
+            csv_file, delimiter=';',
+            quotechar='"', quoting=csv.QUOTE_MINIMAL)
+        writer.writerow(
+            [relative_path.as_posix(), len(result[1]), len(result[2]), len(result[3]),
+             result[4], result[1], result[2], result[3]]
+        )
 
-def run_VarMiddle(java_file):
+
+def execute_python_code_in_parallel_thread(file):
     """ This runs in a separate thread. """
-    # print(java_file)
-    lines = VarMiddle().value(java_file)
-    # print(1)
-    nested_blocks = NestedBlocks(2).value(java_file)
-    # print(6)
-    # lines = pattern.value(java_file)
-    # print('Variables are declared in the middle there: {}'.format(lines))
-    return java_file, lines, nested_blocks
+    var_numbers = VarMiddle().value(file)
+    nested_for_blocks = NestedBlocks(2, block_type=BlockType.FOR).value(file)
+    nested_if_blocks = NestedBlocks(2, block_type=BlockType.IF).value(file)
+    entropy = Entropy().value(file)
+    return file, var_numbers, nested_for_blocks, nested_if_blocks, entropy
 
 
 if __name__ == '__main__':
-    t1 = time.time()
-    pool = Pool(multiprocessing.cpu_count())
+    start = time.time()
+
+    path = '04'
+    os.makedirs(path, exist_ok=True)
+    filename = '04/04-find-patterns.csv'
+    with open(filename, 'w', newline='\n', encoding='utf-8') as csv_file:
+        writer = csv.writer(
+            csv_file, delimiter=';',
+            quotechar='"', quoting=csv.QUOTE_MINIMAL)
+        writer.writerow([
+            'filename', 'var_middle_number', 'nested_for_number',
+            'nested_if_number', 'entropy', 'lines_for_var_middle',
+            'lines_for_cycle', 'lines_for_if']
+        )
+
+    pool = Pool(20)
     handled_files = []
-
+    log_func = partial(log_result, file_to_write=filename)
     with open(args.filename, 'r') as f:
         for i in f.readlines():
             java_file = str(Path(dir_path, i)).strip()
             pool.apply_async(
-                run_VarMiddle,
+                execute_python_code_in_parallel_thread,
                 args=(java_file,),
-                callback=log_result)
+                callback=log_func)
 
     pool.close()
     pool.join()
 
-    t2 = time.time()
-    print('It took this many seconds ' + str(t2 - t1))
+    end = time.time()
+    print('It took {} seconds'.format(str(end - start)))
```

Comment:
@lyriccoder we don't need `str` here anymore, `format` will take care of string conversion.